File: packages/pulse/swift/Sources/Pulse/Endpoint.swift
import Foundation

/// Effects the core emits for the adapter to carry out. See spec §3.
public enum Effect: Equatable {
    /// Send these bytes as ONE message on the current link.
    case transmit([UInt8])
    /// Hand this payload to the application — in order, exactly once.
    /// Hand this payload to the application — in order, exactly once. `durable`
    /// echoes the sender's per-message durable flag so a bridging app can
    /// preserve the intent when forwarding onto another hop. Pure transport info.
    case deliver(seq: UInt64, payload: [UInt8], durable: Bool)
    /// Begin establishing the link (dial).
    case open
    /// Tear down the current link (dead/stale).
    case close
    /// Inbound history before `fromSeq` is unrecoverable; re-sync there.
    case resetInbound(fromSeq: UInt64, peerEpoch: String)
    /// The peer confirmed receipt of every outbound message with seq ≤ `seqUpTo`
    /// (our outbox pruned up to here). Lets the app resolve "delivered" for what
    /// it sent — e.g. clear/roll back optimistic UI. Observational only.
    case acked(seqUpTo: UInt64)
    /// Persist this outbox entry to durable storage (survives a process
    /// restart). Emitted only by a durable-supported endpoint, only for durable
    /// sends. Carries ONLY seq and bytes — never a destination or routing hint.
    case store(seq: UInt64, payload: [UInt8])
    /// Durable entries with seq ≤ `seqUpTo` are confirmed delivered (or expired)
    /// and may be deleted from durable storage.
    case unstore(seqUpTo: UInt64)
}

public enum LinkState: Equatable {
    case disconnected
    case connected
}

/// Per-endpoint durability capability (advertised at handshake). See spec §8.1.
public struct DurableConfig {
    public var supported: Bool
    public var maxRetentionMs: Int
    public init(supported: Bool, maxRetentionMs: Int = 0) {
        self.supported = supported
        self.maxRetentionMs = maxRetentionMs
    }
}

/// Tunable parameters. Defaults in spec §8.
public struct PulseParams {
    public var heartbeatIntervalMs: Int
    public var deadAfterMs: Int
    public var reconnectBaseMs: Int
    public var reconnectMaxMs: Int
    public var reconnectFactor: Double

    public init(
        heartbeatIntervalMs: Int = 15_000,
        deadAfterMs: Int = 30_000,
        reconnectBaseMs: Int = 1_000,
        reconnectMaxMs: Int = 30_000,
        reconnectFactor: Double = 2
    ) {
        self.heartbeatIntervalMs = heartbeatIntervalMs
        self.deadAfterMs = deadAfterMs
        self.reconnectBaseMs = reconnectBaseMs
        self.reconnectMaxMs = reconnectMaxMs
        self.reconnectFactor = reconnectFactor
    }
}

/// Durable snapshot for restart-durability (spec §10).
public struct Snapshot: Equatable {
    public var epoch: String
    public var sendSeq: UInt64
    public var outboxBase: UInt64
    public var outbox: [(seq: UInt64, payload: [UInt8], durable: Bool, sentAt: Int)]
    public var recvCursor: UInt64
    public var peerEpoch: String

    public static func == (l: Snapshot, r: Snapshot) -> Bool {
        l.epoch == r.epoch && l.sendSeq == r.sendSeq && l.outboxBase == r.outboxBase
            && l.recvCursor == r.recvCursor && l.peerEpoch == r.peerEpoch
            && l.outbox.count == r.outbox.count
            && zip(l.outbox, r.outbox).allSatisfy {
                $0.seq == $1.seq && $0.payload == $1.payload && $0.durable == $1.durable
            }
    }
}

/// Endpoint — the sans-I/O core state machine. See spec §2–§7.
///
/// Symmetric and full-duplex: simultaneously producer and consumer. Performs
/// NO I/O — inputs in, `Effect`s out. Deterministic given the same inputs,
/// clock ticks, and injected `random`.
public final class Endpoint {
    private let params: PulseParams
    private let random: () -> Double

    private var epoch: String
    private var sendSeq: UInt64 = 0
    private var outbox: [(seq: UInt64, payload: [UInt8], durable: Bool, sentAt: Int)] = []
    private var outboxBase: UInt64 = 0

    private var recvCursor: UInt64 = 0
    private var peerEpoch: String = ""

    private let durable: DurableConfig
    private var peerDurableSupported = false

    private var state: LinkState = .disconnected
    private var lastRecvAt: Int = 0
    private var lastSendAt: Int = 0
    private var reconnectAt: Int?
    private var attempt: Int = 0
    private var clock: Int = 0

    public init(
        epoch: String,
        params: PulseParams = PulseParams(),
        random: @escaping () -> Double = { Double.random(in: 0..<1) },
        restore: Snapshot? = nil,
        durable: DurableConfig? = nil
    ) {
        self.params = params
        self.random = random
        self.epoch = epoch
        self.durable = durable ?? DurableConfig(supported: false)
        if let s = restore { load(s) }
    }

    // MARK: - Inputs

    /// Application wants to send an opaque payload. Returns assigned seq + effects.
    public func send(_ payload: [UInt8], durable durableFlag: Bool = false) -> (seq: UInt64, effects: [Effect]) {
        sendSeq += 1
        let seq = sendSeq
        // Durable only if the app asked AND we can persist (spec §8.1).
        let isDurable = durableFlag && durable.supported
        // Outbox entry created BEFORE any transmit (spec §3 ordering rule).
        outbox.append((seq: seq, payload: payload, durable: isDurable, sentAt: clock))
        var effects: [Effect] = []
        if isDurable { effects.append(.store(seq: seq, payload: payload)) }
        if state == .connected {
            let wireDurable = durableFlag && peerDurableSupported
            effects.append(transmit(.data(seq: seq, ack: recvCursor, payload: payload, durable: wireDurable)))
        }
        return (seq, effects)
    }

    public func onConnected(_ now: Int) -> [Effect] {
        clock = now
        state = .connected
        attempt = 0
        reconnectAt = nil
        lastRecvAt = now
        var effects: [Effect] = []
        effects.append(
            transmit(
                .hello(
                    epoch: epoch, recvEpoch: peerEpoch, recvCursor: recvCursor,
                    durableSupported: durable.supported, maxRetentionMs: UInt64(durable.maxRetentionMs)),
                now: now))
        return effects
    }

    public func onDisconnected(_ now: Int) -> [Effect] {
        clock = now
        state = .disconnected
        attempt += 1
        reconnectAt = now + backoffDelay(attempt)
        return []
    }

    public func onBytes(_ bytes: [UInt8], _ now: Int) -> [Effect] {
        clock = now
        guard let frame = decodeFrame(bytes) else { return [] }  // malformed ⇒ ignore
        lastRecvAt = now
        switch frame {
        case let .hello(epoch, recvEpoch, recvCursor, durableSupported, _):
            return onHello(
                epoch: epoch, recvEpoch: recvEpoch, recvCursor: recvCursor,
                durableSupported: durableSupported, now: now)
        case let .data(seq, ack, payload, durable):
            return onData(seq: seq, ack: ack, payload: payload, durable: durable, now: now)
        case let .ack(ack):
            return onPeerCursor(ack, now)
        case let .reset(epoch, oldest):
            return onReset(epoch: epoch, oldest: oldest)
        case let .heartbeat(ack):
            return onPeerCursor(ack, now)
        }
    }

    public func onTick(_ now: Int) -> [Effect] {
        clock = now
        var effects: [Effect] = []
        expireDurable(now, &effects)
        if state == .connected {
            if now - lastSendAt >= params.heartbeatIntervalMs {
                effects.append(transmit(.heartbeat(ack: recvCursor), now: now))
            }
            if now - lastRecvAt >= params.deadAfterMs {
                effects.append(.close)
            }
        } else if let at = reconnectAt, now >= at {
            reconnectAt = nil
            effects.append(.open)
        }
        return effects
    }

    /// Drop durable outbox entries past the retention window; emit unstore.
    private func expireDurable(_ now: Int, _ effects: inout [Effect]) {
        let ttl = durable.maxRetentionMs
        if !durable.supported || ttl <= 0 { return }
        let expired = outbox.filter { $0.durable && now - $0.sentAt >= ttl }
        if expired.isEmpty { return }
        let expiredSeqs = Set(expired.map { $0.seq })
        outbox.removeAll { expiredSeqs.contains($0.seq) }
        let highest = expired.map { $0.seq }.max() ?? 0
        effects.append(.unstore(seqUpTo: highest))
    }

    // MARK: - Frame handlers

    private func onHello(
        epoch: String, recvEpoch: String, recvCursor: UInt64, durableSupported: Bool, now: Int
    ) -> [Effect] {
        var effects: [Effect] = []
        peerEpoch = epoch
        peerDurableSupported = durableSupported

        // (a) Peer resuming against an epoch we no longer have.
        if !recvEpoch.isEmpty, recvEpoch != self.epoch {
            effects.append(transmit(.reset(epoch: self.epoch, oldest: outboxBase + 1), now: now))
            resendFrom(outboxBase + 1, &effects, now)
            return effects
        }

        // (b) Prune what the peer already has, then resend the rest — announcing
        // any gap at the head of our outbox (e.g. non-durable lost in a restart).
        if recvCursor >= outboxBase {
            pruneOutbox(recvCursor, &effects)
            resendWithGapAnnounce(recvCursor + 1, &effects, now)
        } else {
            effects.append(transmit(.reset(epoch: self.epoch, oldest: outboxBase + 1), now: now))
            resendFrom(outboxBase + 1, &effects, now)
        }
        return effects
    }

    private func onData(seq: UInt64, ack: UInt64, payload: [UInt8], durable: Bool, now: Int) -> [Effect] {
        var effects: [Effect] = []
        pruneOutbox(ack, &effects)  // peer piggybacks its receipt of our outbound
        if seq == recvCursor + 1 {
            recvCursor = seq
            effects.append(.deliver(seq: seq, payload: payload, durable: durable))
        } else if seq <= recvCursor {
            // Duplicate (a resend because our earlier ack was lost). Re-advertise
            // our cursor so the sender learns we have it and stops resending —
            // without this a lost ack can wedge the sender resending forever and
            // it never observes delivery. (Same rationale as TCP's dup-ACK.)
            effects.append(transmit(.ack(ack: recvCursor), now: now))
        } else {
            // hole: seq > recvCursor+1. Ask peer to rewind.
            effects.append(transmit(.ack(ack: recvCursor), now: now))
        }
        return effects
    }

    /// A peer advertised its receive cursor (explicit ACK or idle HEARTBEAT).
    /// Prune what it confirms; if it lags our latest send, resend the gap so
    /// tail-loss and holes self-heal without a reconnect.
    private func onPeerCursor(_ peerCursor: UInt64, _ now: Int) -> [Effect] {
        var effects: [Effect] = []
        pruneOutbox(peerCursor, &effects)
        if peerCursor < sendSeq {
            resendWithGapAnnounce(peerCursor + 1, &effects, now)
        }
        return effects
    }

    private func onReset(epoch: String, oldest: UInt64) -> [Effect] {
        peerEpoch = epoch
        if oldest > recvCursor + 1 {
            // Unavoidable gap: (recvCursor+1 .. oldest-1) are gone forever.
            recvCursor = oldest - 1
            return [.resetInbound(fromSeq: oldest, peerEpoch: epoch)]
        }
        return []
    }

    // MARK: - Helpers

    private func resendFrom(_ fromSeq: UInt64, _ effects: inout [Effect], _ now: Int) {
        for e in outbox where e.seq >= fromSeq {
            let wireDurable = e.durable && peerDurableSupported
            effects.append(
                transmit(.data(seq: e.seq, ack: recvCursor, payload: e.payload, durable: wireDurable), now: now))
        }
    }

    /// Resend from `fromSeq`, first announcing (via RESET) any gap at the head:
    /// if our oldest retained seq is beyond `fromSeq`, those seqs were discarded
    /// (e.g. non-durable lost in a restart) and can never be filled — without the
    /// RESET the peer treats the resend as a hole, re-ACKs, and we livelock.
    private func resendWithGapAnnounce(_ fromSeq: UInt64, _ effects: inout [Effect], _ now: Int) {
        if let oldest = oldestRetainedSeq(), oldest > fromSeq {
            effects.append(transmit(.reset(epoch: epoch, oldest: oldest), now: now))
        }
        resendFrom(fromSeq, &effects, now)
    }

    private func oldestRetainedSeq() -> UInt64? {
        outbox.map { $0.seq }.min()
    }

    private func pruneOutbox(_ ackSeq: UInt64, _ effects: inout [Effect]) {
        if ackSeq <= outboxBase { return }
        let hadDurable = outbox.contains { $0.seq <= ackSeq && $0.durable }
        outbox.removeAll { $0.seq <= ackSeq }
        outboxBase = ackSeq
        // Surface the confirmed delivery floor so the app can resolve/roll back
        // optimistic UI for messages it sent. Observational only.
        effects.append(.acked(seqUpTo: ackSeq))
        if hadDurable { effects.append(.unstore(seqUpTo: ackSeq)) }
    }

    private func transmit(_ frame: Frame, now: Int? = nil) -> Effect {
        if let n = now { lastSendAt = n }
        // encodeFrame only throws on >255-byte epoch; epochs are bounded by the
        // caller, so a failure here is a programming error — trap it.
        return .transmit(try! encodeFrame(frame))
    }

    private func backoffDelay(_ attempt: Int) -> Int {
        let grown = Double(params.reconnectBaseMs)
            * pow(params.reconnectFactor, Double(attempt - 1))
        let ceil = min(Double(params.reconnectMaxMs), grown)
        return Int(random() * (ceil + 1))  // full jitter: uniform [0, ceil]
    }

    // MARK: - Observation

    public func nextDeadline() -> Int? {
        if state == .connected {
            return min(lastSendAt + params.heartbeatIntervalMs, lastRecvAt + params.deadAfterMs)
        }
        return reconnectAt
    }

    public var link: LinkState { state }
    public var sendSeqValue: UInt64 { sendSeq }
    public var recvCursorValue: UInt64 { recvCursor }
    public var outboxSize: Int { outbox.count }

    public func snapshot() -> Snapshot {
        Snapshot(
            epoch: epoch, sendSeq: sendSeq, outboxBase: outboxBase,
            outbox: outbox, recvCursor: recvCursor, peerEpoch: peerEpoch)
    }

    private func load(_ s: Snapshot) {
        epoch = s.epoch
        sendSeq = s.sendSeq
        outboxBase = s.outboxBase
        outbox = s.outbox
        recvCursor = s.recvCursor
        peerEpoch = s.peerEpoch
    }
}

File: packages/pulse/ts/src/endpoint.ts
/**
 * Endpoint — the sans-I/O core state machine. See spec/PROTOCOL.md §2–§7.
 *
 * Symmetric and full-duplex: every endpoint is simultaneously producer and
 * consumer. Performs NO I/O — inputs in, {@link Effect}s out. Deterministic
 * given the same inputs, clock ticks, and injected `random`.
 */

import {
  DEFAULT_PARAMS,
  type DurableConfig,
  type Effect,
  type EndpointOptions,
  LinkState,
  type Payload,
  type PulseParams,
  type Seq,
  type Snapshot,
} from './types.js';
import { decodeFrame, encodeFrame, type Frame } from './wire.js';

interface OutboxEntry {
  seq: Seq;
  payload: Payload;
  /** Sent with durable:true — persist across restart (only if we support it). */
  durable: boolean;
  /** When first assigned a send time, for retention expiry (ms). */
  sentAt: number;
}

// Isomorphic base64 for snapshot payloads (runs in Node AND the browser). The
// core must not depend on Buffer — a browser caller imports this too.
function b64encode(u: Uint8Array): string {
  let s = '';
  for (let i = 0; i < u.length; i++) s += String.fromCharCode(u[i]!);
  // btoa in the browser; Buffer fallback only if btoa is absent (old Node).
  return typeof btoa === 'function' ? btoa(s) : Buffer.from(u).toString('base64');
}
function b64decode(s: string): Uint8Array {
  if (typeof atob === 'function') {
    const bin = atob(s);
    const out = new Uint8Array(bin.length);
    for (let i = 0; i < bin.length; i++) out[i] = bin.charCodeAt(i);
    return out;
  }
  return new Uint8Array(Buffer.from(s, 'base64'));
}

export class Endpoint {
  private readonly params: PulseParams;
  private readonly random: () => number;

  private epoch: string;
  private sendSeq: Seq = 0n;
  private outbox: OutboxEntry[] = [];
  private outboxBase: Seq = 0n; // (lowest retained seq) - 1

  private recvCursor: Seq = 0n;
  private peerEpoch = '';

  /** My durability capability (advertised in my HELLO). */
  private readonly durable: DurableConfig;
  /** Whether the peer advertised it can persist (learned from its HELLO). */
  private peerDurableSupported = false;

  private state: LinkState = LinkState.Disconnected;
  private lastRecvAt = 0;
  private lastSendAt = 0;
  private reconnectAt: number | null = null;
  private attempt = 0;
  /** Last-known clock (ms), updated on every timed input. Used to stamp sentAt
   *  on send() which has no `now` of its own. */
  private clock = 0;

  constructor(opts: EndpointOptions) {
    this.params = { ...DEFAULT_PARAMS, ...(opts.params ?? {}) };
    this.random = opts.random ?? Math.random;
    this.epoch = opts.epoch;
    this.durable = opts.durable ?? { supported: false };
    if (opts.restore) this.loadSnapshot(opts.restore);
  }

  // ── Inputs ──────────────────────────────────────────────────────────────

  send(payload: Payload, opts?: { durable?: boolean }): { seq: Seq; effects: Effect[] } {
    this.sendSeq += 1n;
    const seq = this.sendSeq;
    // A message is durable only if the app asked AND we can persist. If the app
    // asked but we can't, it degrades to a normal in-memory entry (spec §8.1).
    const durable = opts?.durable === true && this.durable.supported;
    // Outbox entry created BEFORE any transmit (spec §3 ordering rule): the
    // payload is resendable before it is ever entrusted to the wire.
    this.outbox.push({ seq, payload, durable, sentAt: this.clock });
    const effects: Effect[] = [];
    // Persist to durable storage immediately (before transmit), so it survives a
    // restart even if the socket is down right now. Only seq+bytes — no target.
    if (durable) effects.push({ t: 'store', seq, payload });
    if (this.state === LinkState.Connected) {
      // The DATA durable bit is set only if the PEER can persist it; otherwise
      // it's pointless on the wire. (Our own `durable` above governs OUR outbox
      // persistence; this bit tells the peer to persist on ITS side if it's the
      // one that will hold the message onward — e.g. a store-and-forward node.)
      const wireDurable = opts?.durable === true && this.peerDurableSupported;
      effects.push(this.transmit({ t: 'data', seq, ack: this.recvCursor, payload, durable: wireDurable }));
    }
    return { seq, effects };
  }

  onConnected(now: number): Effect[] {
    this.clock = now;
    this.state = LinkState.Connected;
    this.attempt = 0;
    this.reconnectAt = null;
    this.lastRecvAt = now; // give the fresh link a full dead-window grace
    const effects: Effect[] = [];
    effects.push(
      this.transmit(
        {
          t: 'hello',
          epoch: this.epoch,
          recvEpoch: this.peerEpoch,
          recvCursor: this.recvCursor,
          durableSupported: this.durable.supported,
          maxRetentionMs: BigInt(this.durable.maxRetentionMs ?? 0),
        },
        now,
      ),
    );
    return effects;
  }

  onDisconnected(now: number): Effect[] {
    this.clock = now;
    this.state = LinkState.Disconnected;
    this.attempt += 1;
    this.reconnectAt = now + this.backoffDelay(this.attempt);
    return [];
  }

  onBytes(bytes: Uint8Array, now: number): Effect[] {
    this.clock = now;
    const frame = decodeFrame(bytes);
    if (frame === null) return []; // malformed ⇒ ignore (spec §5.0)
    this.lastRecvAt = now;
    switch (frame.t) {
      case 'hello':
        return this.onHello(frame, now);
      case 'data':
        return this.onData(frame, now);
      case 'ack':
        // An explicit ACK is a consumer signaling its cursor (often a hole).
        // Prune what it has, then resend anything it is missing.
        return this.onPeerCursor(frame.ack, now);
      case 'reset':
        return this.onReset(frame);
      case 'heartbeat':
        // An idle heartbeat reveals the consumer's cursor; if it lags our
        // sendSeq (tail-loss), resend the gap. This is what heals a tail lost
        // right before the producer went quiet.
        return this.onPeerCursor(frame.ack, now);
    }
  }

  onTick(now: number): Effect[] {
    this.clock = now;
    const effects: Effect[] = [];
    // Expire durable outbox entries older than our retention window: drop them
    // and tell the adapter to delete them from disk. They will never be resent.
    this.expireDurable(now, effects);
    if (this.state === LinkState.Connected) {
      if (now - this.lastSendAt >= this.params.heartbeatIntervalMs) {
        effects.push(this.transmit({ t: 'heartbeat', ack: this.recvCursor }, now));
      }
      if (now - this.lastRecvAt >= this.params.deadAfterMs) {
        effects.push({ t: 'close' });
      }
    } else if (this.reconnectAt !== null && now >= this.reconnectAt) {
      this.reconnectAt = null;
      effects.push({ t: 'open' });
    }
    return effects;
  }

  /** Drop durable outbox entries past the retention window; emit unstore for
   *  each so the adapter clears disk. Only runs when we have a finite retention
   *  and are the durable-supported side. */
  private expireDurable(now: number, effects: Effect[]): void {
    const ttl = this.durable.maxRetentionMs ?? 0;
    if (!this.durable.supported || ttl <= 0) return;
    const expired = this.outbox.filter((e) => e.durable && now - e.sentAt >= ttl);
    if (expired.length === 0) return;
    // Remove expired entries. unstore floor = highest expired seq that is also
    // contiguous from outboxBase is not required; we emit a precise unstore per
    // the highest expired seq (adapter deletes ≤ that among durable ids it holds).
    const expiredSeqs = new Set(expired.map((e) => e.seq));
    this.outbox = this.outbox.filter((e) => !expiredSeqs.has(e.seq));
    const highest = expired.reduce((m, e) => (e.seq > m ? e.seq : m), 0n);
    effects.push({ t: 'unstore', seqUpTo: highest });
  }

  // ── Frame handlers ────────────────────────────────────────────────────────

  private onHello(
    f: Extract<Frame, { t: 'hello' }>,
    now: number,
  ): Effect[] {
    const effects: Effect[] = [];
    this.peerEpoch = f.epoch;
    // Learn whether the peer can persist — governs the wire durable bit we set.
    this.peerDurableSupported = f.durableSupported;

    // (a) Peer resuming against an epoch we no longer have (we cold-started).
    if (f.recvEpoch !== '' && f.recvEpoch !== this.epoch) {
      effects.push(
        this.transmit({ t: 'reset', epoch: this.epoch, oldest: this.outboxBase + 1n }, now),
      );
      this.resendFrom(this.outboxBase + 1n, effects, now);
      return effects;
    }

    // (b) Prune what the peer already has, then resend the rest — announcing any
    // gap at the head of our outbox (e.g. a non-durable entry lost in a restart).
    if (f.recvCursor >= this.outboxBase) {
      this.pruneOutbox(f.recvCursor, effects);
      this.resendWithGapAnnounce(f.recvCursor + 1n, effects, now);
    } else {
      // Peer is behind our oldest retained seq — we pruned what it needs.
      effects.push(
        this.transmit({ t: 'reset', epoch: this.epoch, oldest: this.outboxBase + 1n }, now),
      );
      this.resendFrom(this.outboxBase + 1n, effects, now);
    }
    return effects;
  }

  private onData(f: Extract<Frame, { t: 'data' }>, now: number): Effect[] {
    const effects: Effect[] = [];
    this.pruneOutbox(f.ack, effects); // peer piggybacks its receipt of our outbound
    if (f.seq === this.recvCursor + 1n) {
      this.recvCursor = f.seq;
      effects.push({ t: 'deliver', seq: f.seq, payload: f.payload, durable: f.durable });
    } else if (f.seq <= this.recvCursor) {
      // Duplicate (a resend because our earlier ack was lost). Re-advertise our
      // cursor so the sender learns we already have it and stops resending —
      // without this, a lost ack can wedge the sender resending forever and it
      // never observes delivery. (Same rationale as TCP's dup-ACK.)
      effects.push(this.transmit({ t: 'ack', ack: this.recvCursor }, now));
    } else {
      // hole: seq > recvCursor+1. Do not deliver; ask peer to rewind.
      effects.push(this.transmit({ t: 'ack', ack: this.recvCursor }, now));
    }
    return effects;
  }

  /**
   * A peer advertised its receive cursor (via explicit ACK or idle HEARTBEAT).
   * Prune what it confirms, and if it is behind our latest send, resend the gap
   * so tail-loss and holes self-heal without a reconnect.
   */
  private onPeerCursor(peerCursor: Seq, now: number): Effect[] {
    const effects: Effect[] = [];
    this.pruneOutbox(peerCursor, effects);
    if (peerCursor < this.sendSeq) {
      this.resendWithGapAnnounce(peerCursor + 1n, effects, now);
    }
    return effects;
  }

  /** Resend outbox entries from `fromSeq`, but first announce (via RESET) any
   *  gap at the head: if our oldest retained seq is beyond `fromSeq`, we can
   *  never fill `fromSeq..oldest-1` (they were discarded — e.g. a non-durable
   *  entry lost in a restart). Without the RESET the peer treats the resend as a
   *  hole, re-ACKs, and we livelock resending forever. */
  private resendWithGapAnnounce(fromSeq: Seq, effects: Effect[], now: number): void {
    const oldest = this.oldestRetainedSeq();
    if (oldest !== null && oldest > fromSeq) {
      effects.push(this.transmit({ t: 'reset', epoch: this.epoch, oldest }, now));
    }
    this.resendFrom(fromSeq, effects, now);
  }

  /** Lowest seq still held in the outbox, or null if empty. */
  private oldestRetainedSeq(): Seq | null {
    let min: Seq | null = null;
    for (const e of this.outbox) {
      if (min === null || e.seq < min) min = e.seq;
    }
    return min;
  }

  private onReset(f: Extract<Frame, { t: 'reset' }>): Effect[] {
    this.peerEpoch = f.epoch;
    if (f.oldest > this.recvCursor + 1n) {
      // Unavoidable gap: (recvCursor+1 .. oldest-1) are gone forever.
      this.recvCursor = f.oldest - 1n;
      return [{ t: 'reset-inbound', fromSeq: f.oldest, peerEpoch: f.epoch }];
    }
    // else: no gap; peer will resend from recvCursor+1 as usual.
    return [];
  }

  // ── Helpers ────────────────────────────────────────────────────────────────

  private resendFrom(fromSeq: Seq, effects: Effect[], now: number): void {
    for (const e of this.outbox) {
      if (e.seq >= fromSeq) {
        // Preserve durable intent on resend: the wire bit reflects whether the
        // peer can persist, matching the original send.
        const wireDurable = e.durable && this.peerDurableSupported;
        effects.push(
          this.transmit(
            { t: 'data', seq: e.seq, ack: this.recvCursor, payload: e.payload, durable: wireDurable },
            now,
          ),
        );
      }
    }
  }

  private pruneOutbox(ackSeq: Seq, effects?: Effect[]): void {
    if (ackSeq <= this.outboxBase) return;
    // Any durable entries being pruned are now confirmed delivered — tell the
    // adapter it may delete them from disk.
    const hadDurable = this.outbox.some((e) => e.seq <= ackSeq && e.durable);
    this.outbox = this.outbox.filter((e) => e.seq > ackSeq);
    this.outboxBase = ackSeq;
    // Surface the confirmed delivery floor so the app can resolve/roll back
    // optimistic UI for messages it sent. Observational only.
    effects?.push({ t: 'acked', seqUpTo: ackSeq });
    if (hadDurable) effects?.push({ t: 'unstore', seqUpTo: ackSeq });
  }

  /** Build a transmit effect and mark send activity. `now` optional for the
   *  data-send path where lastSendAt is refreshed by the caller context. */
  private transmit(frame: Frame, now?: number): Effect {
    if (now !== undefined) this.lastSendAt = now;
    return { t: 'transmit', bytes: encodeFrame(frame) };
  }

  private backoffDelay(attempt: number): number {
    const ceil = Math.min(
      this.params.reconnectMaxMs,
      this.params.reconnectBaseMs * this.params.reconnectFactor ** (attempt - 1),
    );
    return Math.floor(this.random() * (ceil + 1)); // full jitter: uniform [0, ceil]
  }

  // ── Observation ────────────────────────────────────────────────────────────

  nextDeadline(): number | null {
    if (this.state === LinkState.Connected) {
      // Earliest of: next heartbeat due, next dead-check due.
      return Math.min(
        this.lastSendAt + this.params.heartbeatIntervalMs,
        this.lastRecvAt + this.params.deadAfterMs,
      );
    }
    return this.reconnectAt;
  }

  get link(): LinkState {
    return this.state;
  }
  get sendSeqValue(): Seq {
    return this.sendSeq;
  }
  get recvCursorValue(): Seq {
    return this.recvCursor;
  }
  get outboxSize(): number {
    return this.outbox.length;
  }

  snapshot(): Snapshot {
    return {
      epoch: this.epoch,
      sendSeq: this.sendSeq.toString(),
      outboxBase: this.outboxBase.toString(),
      outbox: this.outbox.map((e) => ({
        seq: e.seq.toString(),
        payloadB64: b64encode(e.payload),
        durable: e.durable,
        sentAt: e.sentAt,
      })),
      recvCursor: this.recvCursor.toString(),
      peerEpoch: this.peerEpoch,
    };
  }

  private loadSnapshot(s: Snapshot): void {
    this.epoch = s.epoch;
    this.sendSeq = BigInt(s.sendSeq);
    this.outboxBase = BigInt(s.outboxBase);
    this.outbox = s.outbox.map((e) => ({
      seq: BigInt(e.seq),
      payload: b64decode(e.payloadB64),
      durable: e.durable ?? false,
      sentAt: e.sentAt ?? 0,
    }));
    this.recvCursor = BigInt(s.recvCursor);
    this.peerEpoch = s.peerEpoch;
  }
}
